src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsCompiler.cs

```diff
@@ -0,0 +1,323 @@
+                {
+                    if (IsDisposing || IsDisposed)
+                    {
+                        return;
+                    }
+
+                    if (token.IsCancellationRequested)
+                    {
+                        LogTelemetry(cancelled: true);
+                        return;
+                    }
+
+                    // Grab the next file to compile off the queue
+                    (string fileName, bool ignoreFileWriteTime) = _filesToCompile.FirstOrDefault();
+                    if (fileName == null)
+                    {
+                        break;
+                    }
+
+                    string outputFileName = GetOutputFileName(fileName);
+                    try
+                    {
+                        if (await CompileDesignTimeInputAsync(accessor.Context, fileName, outputFileName, ignoreFileWriteTime, token))
+                        {
+                            compileCount++;
+                        }
+                        // We remove the file if the compilation wasn't cancelled, regardless of whether it really happened or not.
+                        ImmutableInterlocked.TryRemove(ref _filesToCompile, fileName, out _);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        LogTelemetry(cancelled: true);
+                        return;
+                    }
+                }
+
+                LogTelemetry(cancelled: false);
+            });
+
+            void LogTelemetry(bool cancelled)
+            {
+                compileStopWatch.Stop();
+                _telemetryService.PostProperties(TelemetryEventName.TempPEProcessQueue, new[]
+                {
+                    ( TelemetryPropertyName.TempPECompileCount,        (object)compileCount),
+                    ( TelemetryPropertyName.TempPEInitialQueueLength,  initialQueueLength),
+                    ( TelemetryPropertyName.TempPECompileWasCancelled, cancelled),
+                    ( TelemetryPropertyName.TempPECompileDuration,     compileStopWatch.ElapsedMilliseconds)
+                });
+            }
+        }
+
+        /// <summary>
+        /// Gets the XML that describes a TempPE DLL, including building it if necessary
+        /// </summary>
+        /// <param name="fileName">A project relative path to a source file that is a design time input</param>
+        /// <returns>An XML description of the TempPE DLL for the specified file</returns>
+        public Task<string> GetDesignTimeInputXmlAsync(string fileName)
+        {
+            if (_state == null)
+            {
+                throw new InvalidOperationException("Can't get design time input information until project information has been received");
+            }
+
+            // Make sure we're not being asked to compile a random file
+            if (!_state.AllInputs.Contains(fileName, StringComparers.Paths))
+            {
```

Comment:
Is there a race here? By the time I've read the values and passed to them this method it could be out of date and throw?

Reply:
Do you mean that the file could be part of `_state` at the time of the check, and then not part of it when compiling later? That's certainly possible, though it shouldn't result in a throw.

In general with this class I've avoided any locking or synchronization, but instead made things such that they can cope with failure, assuming that failure cases will be rare. That may not pan out to be true, but it's worth noting that this method has no callers at the moment, so I wouldn't bet on its current implementation being 100% correct anyway. I'm doing this is pieces because previously your feedback was that the PRs were too big, too hard to review, and should be split up into components.

Reply:
Where does the caller get a "file name" from? What code they write that lets them call this method without throwing?

Reply:
Ahh I think I see what you mean; If the list of monikers that the consumer of this gets is out of date when this method is called. That does sound possible. I'm also thinking that this might change to simply process the whole queue, rather than pluck out one file, on the assumption that if one file is requested, the rest are likely to follow. 

The thing that will call this, which is the next piece to come, will be ProjectHostBridge-derived, so will be doing the UI thread stuff that does, so as mentioned I wouldn't be surprised if this changes.